README


This archive contains the distribution Config-Model,
version 2.027:

  Create tools to validate, migrate and edit configuration files

This software is Copyright (c) 2012 by Dominique Dumont, Krzysztof Tyszecki.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999




Changes
2.065 2015-01-06

  Bug fixes:
    * Value: avoid undef warning when creating error message
    * Node: Must load override class to be able to use it...

  New (small) feature:
    * Value: warning or error message defined in a model can contain
	  a $_ to show the bad value to user

2.064 2014-12-04

  A small change for this release:
  
    * Version 2.062 added the possibility to override the Perl class
      implementing a configuration node by adding a class parameter in a
      place that is confusing. This release fix this bug: this optional
      override class is now declared at the top of a configuration class.

2.063 2014-11-28

  Bug fix release:
    * GenClassPod: use Exporter instead of Exporter::Lite.
    * adapted contrib/bash_completion for cme-old

2.062 2014-11-23

  Big change for this release:
    * cme command is no longer delivered with this distribution. cme
      is now delivered by App::Cme distribution.

  To help the transition, this distribution delivers cme-old. You can
  use this command until App::Cme is available in CPAN (next few days)
  or if you find bugs in the new cme command (which now uses App::Cmd).

  Other incompatible changes:
	* old config-edit command is now really gone
	* cme extensions are obsolete. Extension mechanism is now provided
	  by App::Cmd

  New feature:
    * The Perl class used to implement node, list and hash
	  configuration element can be overridden using a new 'class'
	  parameter. The Perl class passed to this parameter must inherit
	  Config::Model::Node, or Config::Model::HashId or Config::Model::ListId.

  Bug fixes:
    * Value: skip notification if migrate yields an empty value
    * Model: use logger for tracing legacy translation

2.061 2014-09-23

  Bug fix release:

  * ValueComputer: leave $@ and $! alone in formula.
    Also skip variables like '$ foo'

2.060 2014-08-19

  Bug fix release:

    * value.t: avoid smoke test failures
    * Avoid new perl 5.20 warning (Debian #758320)
    * Value: improved notify change message (computed vs preset vs default)
    * fix pod doc in cme (RT 97605, Debian #756433)
    * depends directly on YAML, not YAML::Any which will be deprecated

2.059 2014-06-29

  New features:

    * cme:
      + added shell command as a shortcut to 'cme edit -ui shell'.
	    E.g 'cme shell ssh' to edit ssh_config through a shell like UI
      + add :@ and :.sort sub command for ordered hash.
        E.g.: "cme modify  dpkg-control ~~ 'binary:~/.*/ Depends:.sort' -save"
        or "cme modify dpkg-copyright ~~ 'Files:.sort' -save "
    * Config::Model::Value: added warn_if parameter

  Bug fixes:

    * cme shell mode:
      * fix or add completion for several commands
      * added shell style pattern match to ll and ls command (e.g 'ls foo*')
    * remove version req from use YAML::Any 0.303 (resolve issues with
      Debian FTBS)
    * Value: fix crash when default value raises a warning and code fix
      returns undef.

2.058 2014-06-19

    Bug fix release

    * cme: modify cryptic user message about "Fixing stuff"
    * Value: tweak migrate_value to always return something, even undef.
      This fixes 'cme check dpkg-copyright' crash (thanks gregoa)
    * Loader: fix pod doc about list operators
    * AnyThing: avoid undef warning that breaks test on freebsd perl 5.10.1

2.057 2014-06-12

  New features:

    * Config::Model: initialise log4perl during construction
    * Value: added file and dir value type. A warning is issued when file
      or dir is not found.

  Other bug fixes:

    * Shortens long index to limit the length of warning and error messages
    * Value: fix check done during apply_fixes...
    * ObjTreeScanner: tolerate deprecated experience parameter.

2.056 2014-05-18

	This release deprecates experience associated with configuration element.
    experience specification in models is now ignored.

	Other Changes:
    * added bash_completion snippet associated to a model. This will be useful
 	  for dpkg-patch model
    * C::M::Lister: skip backup files when listing available applications
	* replace File::Slurp with Path::Tiny in yaml backend test and gen-class-pod
      (RT #95692). (which changes the dependencies)
    * replaced Path::Class with Path::Tiny dependency

2.055 2014-05-02

    This release removes all code related to asynchronous stores.  This code was
    buggy. Making this work correctly would require re-writing most of
    Config::Model.

    * Config::Model::Value:
      * added deprecation warning about callback parameter
      * C::M::Value: removed async code
    * perltidy run on all files
    * removed dependency on namepace::autoclean
    * Removed AnyEvent dependency

2.054 2014-04-01

  Bug fix release:
    * Loader Dumper: fix quote handling (Closes Debian 743097)
    * Loader: return 'ok' after dispatching an action (avoid undef warning
      during tests)
    * cme: -save options force a save even if no semantic change was done
    * ListId: sort may trigger notify_change is elements are actually
      moved around, so 'cme modify stuff list:.sort' will save the file
      as expected.

2.053 2014-03-25

   Bug fix release:
   * Loader: fix broken list leaf assignment (like 'list:4=foo')

2.052 2014-03-23

    This release provides new functionalities to 'cme modify stuff'
    and to the shell like interface ('cme edit -ui shell stuff). See
    Config::Model::Loader doc for details:

      + added remove by value on list and hash ( :-= and :-~ )
        (Closes Debian #741453)
	  + added lots of list and hash operator usable on 'cme modify stuff'.
        E.g. :.insort :.insert_before :.insert_at :.unshift  :.push :.sort
      + handle list/hash value substitution ( e.g. list:=~s/foo/bar/ )
      * warn when trying to remove a list element with a non numeric index
      * rationalize list and hash commands: list assignment is now
        list:=a,b,c instead of list=a,b,c

	Other changes:


Dumper.pm
# $Author: ddumont $
# $Date: 2008-04-03 19:06:58 +0200 (Thu, 03 Apr 2008) $
# $Revision: 581 $

#    Copyright (c) 2006-2007 Dominique Dumont.
#
#    This file is part of Config-Model.
#
#    Config-Model is free software; you can redistribute it and/or
#    modify it under the terms of the GNU Lesser Public License as
#    published by the Free Software Foundation; either version 2.1 of
#    the License, or (at your option) any later version.
#
#    Config-Model is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#    Lesser Public License for more details.
#
#    You should have received a copy of the GNU Lesser Public License
#    along with Config-Model; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA

package Config::Model::Dumper;
use Carp;
use strict;
use warnings ;

use Config::Model::Exception ;
use Config::Model::ObjTreeScanner ;

use vars qw($VERSION);
$VERSION = sprintf "1.%04d", q$Revision: 581 $ =~ /(\d+)/;

=head1 NAME

Config::Model::Dumper - Serialize data of config tree

=head1 SYNOPSIS

 use Config::Model ;

 # create your config model
 my $model = Config::Model -> new ;
 $model->create_config_class( ... ) ;

 # create instance
 my $inst = $model->instance (root_class_name => 'FooBar', 
			      instance_name => 'test1');

 # create root of config
 my $root = $inst -> config_root ;

 # put some data in config tree
 my $step = 'std_id:ab X=Bv - std_id:bc X=Av - a_string="toto tata"';
 $root->walk( step => $step ) ;

 # dump only customized data (audit mode)
 print $root->dump_tree;

 # dump all data including default values
 print $root->dump_tree( full_dump => 1 ) ;

=head1 DESCRIPTION

This module is used directly by L<Config::Model::Node> to serialize 
configuration data in a compact (but readable) string.

The serialisation can be done in standard mode where only customized
values are dumped in the string. I.e. only data modified by the user
are dumped.

The other mode is C<full_dump> mode where all all data, including
default values, are dumped.

The serialized string can be used by L<Config::Model::Walker> to store
the data back into a configuration tree.

=head1 CONSTRUCTOR

=head2 new ( )

No parameter. The constructor should be used only by
L<Config::Model::Node>.

=cut

sub new {
    bless {}, shift ;
}

=head1 Methods

=head2 dump_tree

Return a string that contains a dump of the object tree with all the
values. This string follows the convention defined by
L<Config::Model::Walker>.

The serialized string can be used by L<Config::Model::Walker> to store
the data back into a configuration tree.

Parameters are:

=over

=item mode ( full | preset )

C<full> will dump all configuration data including default
values. 

C<preset> will dump only value entered in preset mode.

By default, the dump contains only data modified by the user
(i.e. data differ from default or preset values).

=item node

Reference to the L<Config::Model::Node> object that is dumped. All
nodes and leaves attached to this node are also dumped.

=item skip_auto_write

Skip node that have a C<cds write> capabality in their model. See
L<Config::Model::AutoRead>.

=item auto_vivify

Scan and create data for nodes elements even if no actual data was
stored in them. This may be useful to trap missing mandatory values.

=back

=cut

sub dump_tree {
    my $self = shift;

    my %args = @_;
    my $full = delete $args{full_dump} || 0;
    my $skip_aw = delete $args{skip_auto_write} || 0 ;
    my $auto_v  = delete $args{auto_vivify}     || 0 ;
    my $mode = delete $args{mode} || '';
    if ($mode and $mode ne 'full' and $mode ne 'preset') {
	croak "dump_tree: unexpected 'mode' value";
    }

    # mode paramter is slightly different from fetch's mode
    my $fetch_mode = $full             ? ''
                   : $mode eq 'full'   ? ''
                   : $mode eq 'preset' ? $mode
                   :                     'custom';

    my $node = delete $args{node} 
      || croak "dump_tree: missing 'node' parameter";

    my $compute_pad = sub {
	my $depth = 0 ;
	my $obj = shift ;
	while (defined $obj->parent) { 
	    $depth ++ ;
	    $obj = $obj->parent ;
	}
        return '  ' x $depth;
    };

    my $std_cb = sub {
        my ( $scanner, $data_r, $obj, $element, $index, $value_obj ) = @_;

	# get value or only customized value
	my $value = $value_obj->fetch ($fetch_mode) ;

	if (    defined $value 
	    and ($value =~ /(\s|")/ or $value eq '')
	   ) {
	    $value =~ s/"/\\"/g ; # escape present quotes
	    $value = '"' . $value . '"' ; # add my quotes
	}

        my $pad = $compute_pad->($obj);

        my $name = defined $index && $index =~ /\s/ ? "$element:\"$index\"" 
	         : defined $index                   ? "$element:$index" 
                 :                                     $element;

        $$data_r .= "\n" . $pad . $name . '=' . $value if defined $value;
    };

    my $check_list_cb = sub {
        my ( $scanner, $data_r, $obj, $element, $index, $value_obj ) = @_;

	# get value or only customized value
	my $value = $value_obj->fetch ($fetch_mode) ;

        my $pad = $compute_pad->($obj);

        my $name = defined $index && $index =~ /\s/ ? "$element:\"$index\"" 
	         : defined $index                   ? "$element:$index" 
                 :                                     $element;

        $$data_r .= "\n" . $pad . $name . '=' . $value if $value;
    };

    my $list_element_cb = sub {
        my ( $scanner, $data_r, $obj, $element, @keys ) = @_;

        my $pad      = $compute_pad->($obj);
	my $list_obj = $obj->fetch_element($element) ;
        my $elt_type = $list_obj->cargo_type ;

        if ( $elt_type eq 'node' ) {
            foreach my $k ( @keys ) {
                $scanner->scan_hash( $data_r, $obj, $element, $k );
            }
        }
        else {
	    no warnings "uninitialized" ;
            $$data_r .= "\n$pad$element=" 
	      . join( ',', $list_obj->fetch_all_values($fetch_mode) ) if @keys;
        }
    };

    my $element_cb = sub {
        my ( $scanner, $data_r, $obj, $element, $key, $next ) = @_;

        my $type = $obj -> element_type($element);

        return if $skip_aw and $next->is_auto_write_for_type('cds') ;

        my $pad = $compute_pad->($obj);
        $$data_r .= "\n$pad$element";
	if ($type eq 'list' or $type eq 'hash') {
	    $key = '"' . $key . '"' if  $key =~ /\s/;
	    $$data_r .= ":$key" ;
	}

        $scanner->scan_node($data_r, $next);
    };

    my @scan_args = (
		     permission      => delete $args{permission} || 'master',
		     fallback        => 'all',
		     auto_vivify     => $auto_v,
		     list_element_cb => $list_element_cb,
		     leaf_cb         => $std_cb,
		     node_element_cb => $element_cb,
		     check_list_element_cb => $check_list_cb,
		     up_cb           => sub { ${$_[1]} .= ' -'; }
		    );

    my @left = keys %args;
    croak "Dumper: unknown parameter:@left" if @left;

    # perform the scan
    my $view_scanner = Config::Model::ObjTreeScanner->new(@scan_args);

    my $ret = '' ;
    $view_scanner->scan_node(\$ret, $node);

    substr( $ret, 0, 1, '' );    # remove leading \n
    return $ret . "\n";
}

1;

=head1 AUTHOR

Dominique Dumont, (ddumont at cpan dot org)

=head1 SEE ALSO

L<Config::Model>,L<Config::Model::Node>,L<Config::Model::Walker>
